Source/bmalloc/libpas/src/libpas/pas_stats.h

```diff
@@ -0,0 +1,311 @@
+
+// FIXME: in principle it should be possible to automatically generate this via PAS_STATS_FOR_EACH_COUNTER, somehow
+#define PAS_RECORD_STAT_malloc_info_bytes(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, size)
+#define PAS_RECORD_STAT_malloc_info_allocations(data, heap_type, size) \
+    pas_stats_malloc_info_record(data, heap_type, size, 1)
+
+#else /* !PAS_ENABLE_STATS */
+#define PAS_RECORD_STAT(name, ...) do { PAS_UNUSED_V(__VA_ARGS__); } while (0)
+#endif /* PAS_ENABLE_STATS */
+
+#define PAS_RECORD_STAT_MALLOC(heap_type, size) \
+    PAS_RECORD_STAT(malloc_info_bytes, heap_type, size); \
+    PAS_RECORD_STAT(malloc_info_allocations, heap_type, size);
+
+/******************************************/
+/* Back to general stat-counter machinery */
+/******************************************/
+
+#if PAS_ENABLE_STATS
+
+void pas_stats_do_accounting_before_recording_stat(void);
+// Assumes g_pas_stats_data.log_lock is held
+void pas_stats_log_all_enabled_stats(void);
+
+// It's OK for the .enabled check here to be non-atomic: we know that all
+// .enabled bits will start initialized to 1, and will be set to 0 at most once.
+// If a thread fails to observe that write-to-0, the consequence is that it will
+// make a few unnecessary atomic writes to some stat-counters, but those stat
+// counters will never actually be used since pas_stats_log_all_enabled_stats
+// does use an atomic check for whether .enabled is set.
+#define PAS_RECORD_STAT(name, ...) do { \
+        if (g_pas_stats_data.per_stat_data. name .base.enabled) { \
+            pas_stats_do_accounting_before_recording_stat(); \
+            PAS_RECORD_STAT_ ## name(&(g_pas_stats_data.per_stat_data. name ), __VA_ARGS__); \
+        } \
+    } while (0)
+
+#define DECLARE_FIELD(name, type, logger) type name;
+typedef struct {
+    PAS_STATS_FOR_EACH_COUNTER(DECLARE_FIELD)
+} pas_stats_per_stat_data;
+#undef DECLARE_FIELD
+
+struct pas_stats_sink {
+    pas_stats_sink_output_func output_func;
+    void* context;
+};
+
+typedef struct {
+    uint64_t log_counter;
+    uint64_t first_log_time_ns;
+    uint64_t pid;
+
+    pas_stats_print_buffer buf;
+
+    pas_stats_sink sink;
+    pas_lock log_lock; // Guards everything except per_stat_data
+    pas_stats_per_stat_data per_stat_data;
+} pas_stats_data;
+
+extern pas_stats_data g_pas_stats_data;
+
+#define PAS_STATS_LOG_COUNT_INTERVAL (1 << 16)
+
+// This setup is only called the first time statistics are actually logged.
+// We use it to handle enablement via a sort of hack: all stats are enabled to
+// begin with, but the first one to accrue a counter will check envvars and
+// disable all the ones which shouldn't actually be enabled. This may lead to
+// some unecessary atomic writes at the beginning, but it avoids placing an
+// extra pthread_once check inline with every statistic, both enabled and
+// disabled.
+void pas_stats_setup_if_necessary(void);
```

Comment:
Since `pas_stats_setup_if_necessary()` is only ever called from `pas_stats_do_accounting_before_recording_stat_slow_path()`, let's make it a static function in the .c file instead.  Is there any reason to declare it publicly like this?